File: hearth/media/js/urls.js
define('urls',
    ['format', 'routes_api', 'routes_api_args', 'settings', 'user', 'utils'],
    function(format, api_endpoints, api_args, settings, user) {

    var group_pattern = /\(.+\)/;
    var optional_pattern = /(\(.*\)|\[.*\]|.)\?/g;
    var reverse = function(view_name, args) {
        args = args || [];
        for (var i in routes) {
            var route = routes[i];
            if (route.view_name != view_name)
                continue;

            // Strip the ^ and $ from the route pattern.
            var url = route.pattern.substring(1, route.pattern.length - 1);

            if (url.indexOf('?') !== -1) {
                url = url.replace(optional_pattern, '');
            }

            // Replace each matched group with a positional formatting placeholder.
            var pos = 0;
            while (group_pattern.test(url)) {
                url = url.replace(group_pattern, '{' + pos++ + '}');
            }

            // Check that we got the right number of arguments.
            if (args.length !== pos) {
                console.error('Expected ' + pos + ' args, got ' + args.length);
                throw new Error('Wrong number of arguments passed to reverse(). View: "' + view_name + '", Argument "' + args + '"');
            }

            return format.format(url, args);

        }
        console.error('Could not find the view "' + view_name + '".');
    };

    function _userArgs(func) {
        return function() {
            var out = func.apply(this, arguments);
            var args = api_args();
            if (user.logged_in()) {
                args._user = user.get_token();
            }
            var blacklist = settings.api_param_blacklist || [];
            for (var key in args) {
                if (!args[key] || blacklist.indexOf(key) !== -1) {
                    delete args[key];
                }
            }
            return require('utils').urlparams(out, args);
        };
    }

    function api(endpoint, args, params) {
        if (!(endpoint in api_endpoints)) {
            console.error('Invalid API endpoint: ' + endpoint);
            return '';
        }
        var url = settings.api_url + format.format(api_endpoints[endpoint], args || []);
        if (params) {
            return require('utils').urlparams(url, params);
        }
        return url;
    }

    function apiParams(endpoint, params) {
        return api(endpoint, [], params);
    }

    function media(path) {
        var media_url = settings.media_url;
        if (media_url[media_url.length - 1] !== '/') {
            media_url += '/';
        }
        if (path[0] === '/') {
            path = path.substr(1);
        }
        return media_url + path;
    }

    return {
        reverse: reverse,
        api: {
            url: _userArgs(api),
            params: _userArgs(apiParams),
            sign: _userArgs(function(url) {return url;}),
            unsigned: {
                url: api,
                params: apiParams
            }
        },
        media: media
    };
});

File: spaceheater/urls.py
import re
from urllib import urlencode
from urlparse import parse_qs

import settings


def strip_ws(data):
    return data.replace('\n', '').replace('\r', '')


API_ROUTES = {}
URL_ROUTES = []

with open('../hearth/media/js/urls.js') as urls_js:
    url_data = urls_js.read()
    arp = re.compile(r'api_endpoints = (\{.+?\});', re.DOTALL)
    arp_data = strip_ws(arp.search(url_data).group(1))
    # shame on me
    API_ROUTES = eval(arp_data)

with open('../hearth/media/js/routes.js') as routes_js:
    url_data = routes_js.read()
    urp = re.compile(r'routes = (\[.+?\]);', re.DOTALL)
    url_data = strip_ws(urp.search(url_data).group(1))
    # shame on me
    URL_ROUTES = eval(url_data)

    # # Convert regex patterns to compiled regexes
    # for route in URL_ROUTES:
    #     route['re'] = re.compile(route['pattern'])


GROUP_PATTERN = re.compile(r'\(.+?\)')


def _qs(url):
    if '?' not in url:
        return url, ''
    qpos = url.index('?')
    return url[:qpos], url[qpos + 1:]


def _urlparams(url, **kw):
    base, qs = _qs(url)

    args = parse_qs(qs)
    args.update(kw)

    return '%s?%s' % (base, urlencode(args))


def _userargs(func):
    def wrap(*args, **kwargs):
        url = func(*args, **kwargs)
        # TODO: Implement this.
        # TODO: Also implement user tokens from cookies.
        url = _urlparams(url, lang='unknown', region='unknown')
        return url
    return wrap


def urlunparam(url, params): 
    base, qs = _qs(url)

    args = parse_qs(qs)
    for param in params:
        if param in args:
            del args[param]

    return '%s?%s' % (base, urlencode(args))


@_userargs
def api_url(path, args=None):
    if not args:
        args = []
    if path not in API_ROUTES:
        raise Exception('API route not found: %s' % path)
    return settings.api_url + API_ROUTES[path].format(*args)


def api_apiParams(path, **kw):
    return _urlparams(api_url(path), **kw)


def reverse(view_name, args=None):
    if not args:
        args = []

    for route in URL_ROUTES:
        if route['view_name'] != view_name:
            continue

        url = route['pattern'][1:-1]  # Strip the first and last chars.
        arg_groups = GROUP_PATTERN.findall(url)
        if not arg_groups:
            return url
        for i, _ in enumerate(arg_groups):
            url = GROUP_PATTERN.sub('{%d}' % i, url)

        if len(args) != len(arg_groups):
            raise Exception('Wrong number of arguments passed to reverse()')

        return url.format(*args)
